pkg/sendbuffer/sendbuffer.go

```diff
@@ -0,0 +1,264 @@
+		maxSendSize:     defaultMaxSendSize,
+		dataKeys:        make([][]byte, 0),
+		sender:          sender,
+		sendInterval:    5 * time.Second,
+		logger:          log.NewNopLogger(),
+		kvStore:         inmemory.NewStore(log.NewNopLogger()),
+		stopSendingChan: make(chan struct{}),
+		isSending:       false,
+	}
+
+	for _, opt := range opts {
+		opt(sb)
+	}
+
+	sb.logger = log.With(sb.logger, "component", "sendbuffer")
+
+	// get all data keys and calculate starting size
+	sb.kvStore.ForEach(func(k, v []byte) error {
+		sb.size += len(k) + len(v)
+		sb.dataKeys = append(sb.dataKeys, k)
+		return nil
+	})
+
+	// sort the keys
+	sort.Slice(sb.dataKeys, func(i, j int) bool {
+		return bytes.Compare(sb.dataKeys[i], sb.dataKeys[j]) < 0
+	})
+
+	return sb
+}
+
+func (sb *SendBuffer) Write(data []byte) (int, error) {
+	sb.writeMutex.Lock()
+	defer sb.writeMutex.Unlock()
+
+	if len(data) == 0 {
+		return 0, nil
+	}
+
+	// if the single data piece is larger than the max send size, drop it and log
+	if len(data) > sb.maxSendSize {
+		sb.logger.Log(
+			"msg", "dropped data because element greater than max send size",
+			"size_of_data", len(data),
+			"max_send_size", sb.maxSendSize,
+			"head", string(data)[0:minInt(len(data), 100)],
+		)
+		return len(data), nil
+	}
+
+	// if we are full, something has backed up
+	// purge everything
+	if len(data)+sb.size > sb.maxSize {
+		sb.logger.Log(
+			"msg", "reached capacity, dropping all data and starting over",
+			"size_of_data", len(data),
+			"buffer_size", sb.size,
+			"size_plus_data", sb.size+len(data),
+			"max_size", sb.maxSize,
+		)
+
+		if err := sb.purge(len(sb.dataKeys)); err != nil {
+			// something has gone horribly wrong
+			sb.logger.Log("msg", "failed to purge", "err", err)
+			return len(data), nil
+		}
+
+		sb.size = 0
+	}
+
+	dataKey := []byte(ulid.New())
+	if err := sb.kvStore.Set(dataKey, data); err != nil {
+		// log it and move on
+		sb.logger.Log("msg", "failed to store data", "err", err)
+		return len(data), nil
+	}
+
+	sb.dataKeys = append(sb.dataKeys, dataKey)
+	sb.size += len(data) + len(dataKey)
+	return len(data), nil
+}
+
+func (sb *SendBuffer) StartSending() {
+	if sb.isSending {
+		return
+	}
+
+	sb.isSending = true
+
+	go func() {
+		ticker := time.NewTicker(sb.sendInterval)
+		defer ticker.Stop()
+
+		for {
+			if err := sb.sendAndPurge(); err != nil {
+				sb.logger.Log("msg", "failed to send and purge", "err", err)
+			}
+
+			select {
+			case <-ticker.C:
+				continue
+			case <-sb.stopSendingChan:
+				sb.isSending = false
+				return
+			}
+		}
+	}()
+}
+
+func (sb *SendBuffer) StopSending() {
+	select {
+	case sb.stopSendingChan <- struct{}{}:
+	default:
+	}
+}
+
+func (sb *SendBuffer) sendAndPurge() error {
+	sb.writeMutex.Lock()
+	defer sb.writeMutex.Unlock()
+
+	if len(sb.dataKeys) == 0 {
+		return nil
+	}
+
+	removeDataKeysToIndex := 0
+	sendBuffer := &bytes.Buffer{}
+	bytesToDelete := 0
+
+	for _, dataKey := range sb.dataKeys {
```

Comment:
The foreach doesn't guarantee any kind of order, this was the driving factor behind tracking the data keys. If were not concerned about order we could drop it.

Reply:
_ah_. That does feel harder. I think that means we'd either end up tracking the data keys (as you do here) or doing an iterate and sort each time. 

I think tracking the data keys brings in another mutex, and some amount of synchronization between them. But using foreach has extra communication. 

I wonder if trying to use kvstore ends up working against this. the ring stuff I was playing with in #1183 is probably a better fit, but I'm probably being overly eager in optimizing. Okay.

Reply:
If we don't care about care about persistance between restarts ... the yea I suppose there is no reason to use the KV store. I chose it would make it easy to add persistance later if we wanted, but for simplicity may we just keep a list? And if we ever want persistance we'll add it later, would not be too hard.

Reply:
dropped KV store